File: src/java/net/percederberg/grammatica/parser/Production.java
/*
 * Production.java
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the BSD license.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * LICENSE.txt file for more details.
 *
 * Copyright (c) 2003-2015 Per Cederberg. All rights reserved.
 */

package net.percederberg.grammatica.parser;

import java.util.ArrayList;

/**
 * A production node. This class represents a grammar production (i.e.
 * a list of child nodes) in a parse tree. The productions are created
 * by a parser, that adds children a according to a set of production
 * patterns (i.e. grammar rules).
 *
 * @author   Per Cederberg
 * @version  1.1
 */
public class Production extends Node {

    /**
     * The production pattern used for this production.
     */
    private ProductionPattern pattern;

    /**
     * The child nodes.
     */
    private ArrayList children;

    /**
     * Creates a new production node.
     *
     * @param pattern        the production pattern
     */
    public Production(ProductionPattern pattern) {
        this.pattern = pattern;
        this.children = new ArrayList();
    }

    /**
     * Checks if this node is hidden, i.e. if it should not be visible
     * outside the parser.
     *
     * @return true if the node should be hidden, or
     *         false otherwise
     */
    boolean isHidden() {
        return pattern.isSynthetic();
    }

    /**
     * Returns the production pattern for this production.
     *
     * @return the production pattern
     */
    public ProductionPattern getPattern() {
        return pattern;
    }

    /**
     * Returns the production (pattern) id. This value is set as a
     * unique identifier when creating the production pattern to
     * simplify later identification.
     *
     * @return the production id
     */
    public int getId() {
        return pattern.getId();
    }

    /**
     * Returns the production node name.
     *
     * @return the production node name
     */
    public String getName() {
        return pattern.getName();
    }

    /**
     * Returns the number of child nodes.
     *
     * @return the number of child nodes
     */
    public int getChildCount() {
        return children.size();
    }

    /**
     * Returns the child node with the specified index.
     *
     * @param index          the child index, starting at 0
     *
     * @return the child node found, or
     *         null if index out of bounds
     */
    public Node getChildAt(int index) {
        if (index < 0 || index >= children.size()) {
            return null;
        } else {
            return (Node) children.get(index);
        }
    }

    /**
     * Adds a child node. The node will be added last in the list of
     * children.
     *
     * @param child          the child node to add
     */
    public void addChild(Node child) {
        if (child != null) {
            child.setParent(this);
            children.add(child);
        }
    }

    /**
     * Returns a string representation of this production.
     *
     * @return a string representation of this production
     */
    public String toString() {
        StringBuffer  buffer = new StringBuffer();

        buffer.append(pattern.getName());
        buffer.append('(');
        buffer.append(pattern.getId());
        buffer.append(')');

        return buffer.toString();
    }
}

File: src/csharp/PerCederberg.Grammatica.Runtime/Production.cs
/*
 * Production.cs
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 * MA 02111-1307, USA.
 *
 * Copyright (c) 2003-2005 Per Cederberg. All rights reserved.
 */

using System.Collections;

namespace PerCederberg.Grammatica.Runtime {

    /**
     * A production node. This class represents a grammar production
     * (i.e. a list of child nodes) in a parse tree. The productions
     * are created by a parser, that adds children a according to a
     * set of production patterns (i.e. grammar rules).
     *
     * @author   Per Cederberg, <per at percederberg dot net>
     * @version  1.5
     */
    public class Production : Node {

        /**
         * The production pattern used for this production.
         */
        private ProductionPattern pattern;

        /**
         * The child nodes.
         */
        private ArrayList children;

        /**
         * Creates a new production node.
         *
         * @param pattern        the production pattern
         */
        public Production(ProductionPattern pattern) {
            this.pattern = pattern;
            this.children = new ArrayList();
        }

        /**
         * The node type id property (read-only). This value is set as
         * a unique identifier for each type of node, in order to
         * simplify later identification.
         *
         * @since 1.5
         */
        public override int Id {
            get {
                return pattern.Id;
            }
        }

        /**
         * The node name property (read-only).
         *
         * @since 1.5
         */
        public override string Name {
            get {
                return pattern.Name;
            }
        }

        /**
         * The child node count property (read-only).
         *
         * @since 1.5
         */
        public override int Count {
            get {
                return children.Count;
            }
        }

        /**
         * The child node index (read-only).
         *
         * @param index          the child index, 0 <= index < Count
         *
         * @return the child node found, or
         *         null if index out of bounds
         *
         * @since 1.5
         */
        public override Node this[int index] {
            get {
                if (index < 0 || index >= children.Count) {
                    return null;
                } else {
                    return (Node) children[index];
                }
            }
        }

        /**
         * Adds a child node. The node will be added last in the list of
         * children.
         *
         * @param child          the child node to add
         */
        public void AddChild(Node child) {
            if (child != null) {
                child.SetParent(this);
                children.Add(child);
            }
        }

        /**
         * The production pattern property (read-only). This property
         * contains the production pattern linked to this production.
         *
         * @since 1.5
         */
        public ProductionPattern Pattern {
            get {
                return pattern;
            }
        }

        /**
         * Returns the production pattern for this production.
         *
         * @return the production pattern
         *
         * @see #Pattern
         *
         * @deprecated Use the Pattern property instead.
         */
        public ProductionPattern GetPattern() {
            return Pattern;
        }

        /**
         * Checks if this node is hidden, i.e. if it should not be visible
         * outside the parser.
         *
         * @return true if the node should be hidden, or
         *         false otherwise
         */
        internal override bool IsHidden() {
            return pattern.Synthetic;
        }

        /**
         * Returns a string representation of this production.
         *
         * @return a string representation of this production
         */
        public override string ToString() {
            return pattern.Name + '(' + pattern.Id + ')';
        }
    }
}
